homeassistant/components/squeezebox/media_player.py

```diff
@@ -65,79 +67,118 @@
+    """Set up an LMS Server from a config entry."""
+    config = config_entry.data
+    _LOGGER.debug("Reached async_setup_entry, config=%s", config)
 
     username = config.get(CONF_USERNAME)
     password = config.get(CONF_PASSWORD)
+    host = config[CONF_HOST]
+    port = config[CONF_PORT]
 
-    if discovery_info is not None:
-        host = discovery_info.get("host")
-        port = discovery_info.get("port")
-    else:
-        host = config.get(CONF_HOST)
-        port = config.get(CONF_PORT)
+    hass.data.setdefault(DOMAIN, {})
 
-    # In case the port is not discovered
-    if port is None:
-        port = DEFAULT_PORT
+    known_players = hass.data[DOMAIN].get(KNOWN_PLAYERS)
+    if known_players is None:
+        hass.data[DOMAIN][KNOWN_PLAYERS] = known_players = []
 
-    # Get IP of host, to prevent duplication of same host (different DNS names)
-    try:
-        ipaddr = await hass.async_add_executor_job(socket.gethostbyname, host)
-    except OSError as error:
-        _LOGGER.error("Could not communicate with %s:%d: %s", host, port, error)
-        raise PlatformNotReady from error
-
-    if ipaddr in known_servers:
-        return
-
-    _LOGGER.debug("Creating LMS object for %s", ipaddr)
+    _LOGGER.debug("Creating LMS object for %s", host)
     lms = Server(async_get_clientsession(hass), host, port, username, password)
-    known_servers.add(ipaddr)
-
-    players = await lms.async_get_players()
-    if players is None:
-        raise PlatformNotReady
-    media_players = []
-    for player in players:
-        media_players.append(SqueezeBoxDevice(player))
-
-    hass.data[DATA_SQUEEZEBOX].extend(media_players)
-    async_add_entities(media_players)
 
+    async def _discovery(now=None):
+        """Discover squeezebox players by polling server."""
+
+        async def _discovered_player(player):
+            """Handle a (re)discovered player."""
+            entity = next(
+                (
+                    known
+                    for known in known_players
+                    if known.unique_id == player.player_id
+                ),
+                None,
+            )
+            if entity and not entity.available:
+                # check if previously unavailable player has connected
+                await player.async_update()
+                entity.available = player.connected
+            if not entity:
+                _LOGGER.debug("Adding new entity: %s", player)
+                entity = SqueezeBoxEntity(player)
+                known_players.append(entity)
+                async_add_entities([entity])
+
+        players = await lms.async_get_players()
+        if players:
+            for player in players:
+                hass.async_create_task(_discovered_player(player))
+
+        hass.helpers.event.async_call_later(DISCOVERY_INTERVAL, _discovery)
+
+    _LOGGER.debug("Adding player discovery job for LMS server: %s", host)
+    hass.async_add_job(_discovery)
```

Comment:
`async_add_job` is legacy and shouldn't be used in new code.

Reply:
Ok. Since _discovery is a coroutine, I used asyncio.create_task instead.